src/mongo/db/pipeline/expression.cpp

```diff
@@ -3072,6 +3085,104 @@ const char* ExpressionWeek::getOpName() const {
     return "$week";
 }
 
+/* ------------------------- ExpressionIsoDayOfWeek --------------------- */
+
+Value ExpressionIsoDayOfWeek::evaluateInternal(Variables* vars) const {
+    Value pDate(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(pDate.coerceToTm()));
+}
+
+int ExpressionIsoDayOfWeek::extract(const tm& tm) {
+    return (tm.tm_wday - 7) % 7 + 7;
+}
+
+REGISTER_EXPRESSION(isoDayOfWeek, ExpressionIsoDayOfWeek::parse);
+const char* ExpressionIsoDayOfWeek::getOpName() const {
+    return "$isoDayOfWeek";
+}
+
+/* ------------------------- ExpressionIsoWeekYear ---------------------- */
+
+Value ExpressionIsoWeekYear::evaluateInternal(Variables* vars) const {
+    Value pDate(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(pDate.coerceToTm()));
+}
+
+int ExpressionIsoWeekYear::extract(const tm& tm) {
+    if (tm.tm_mon > 0 && tm.tm_mon < 11) {
+        return tm.tm_year + 1900;
+    } else if (tm.tm_mon == 0) {
+        int isoWeek = ExpressionIsoWeek::extract(tm);
+        if (isoWeek > 51) return tm.tm_year + 1900 - 1;
+        else return tm.tm_year + 1900;
+    } else {
+        int isoWeek = ExpressionIsoWeek::extract(tm);
+        if (isoWeek == 1) return tm.tm_year + 1900 + 1;
+        else return tm.tm_year + 1900;
+    }
+}
+
+REGISTER_EXPRESSION(isoWeekYear, ExpressionIsoWeekYear::parse);
+const char* ExpressionIsoWeekYear::getOpName() const {
+    return "$isoWeekYear";
+}
+
+/* ------------------------- ExpressionIsoWeek -------------------------- */
+
+Value ExpressionIsoWeek::evaluateInternal(Variables* vars) const {
+    Value pDate(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(pDate.coerceToTm()));
+}
+
+int ExpressionIsoWeek::lastWeek(int year) {
+    struct tm tm;
+    tm.tm_year = year - 1900;
+    tm.tm_mon = 11;
+    tm.tm_mday = 31;
+    tm.tm_hour = 23;
+    tm.tm_min = 59;
+    tm.tm_sec = 59;
+    mktime(&tm);
+    if (tm.tm_wday > 0 && tm.tm_wday < 4) {
```

Comment:
Everything from here down is confusing and non-intuitive to me. After reading the wikipedia article on "ISO week date" (https://en.wikipedia.org/wiki/ISO_week_date) I'm still not following. Can you add comments and/or point me to where you got this logic?